ECAM cabin press page

nikotina91 opened the pull request:
Work in progress for the ecam pressure page, any reference material, feedback or input is appreciated.

## Summary of Changes

- [x] Added all the visual elements of the cabin press page
- [x] "SAFETY" turns amber when safety valve is open
- [x] Psi delta value turns amber when > 8.5psi or < -0.4psi
- [x] V/S value turns amber when > 2000ft/m or < -2000ft/m
- [x]  cab alt value turns red when > 9550ft
- [x] pack 1 and 2 triangles + text turn amber when corresponding pack flow valve is closed with the related engine running
- [x] all three gauges work
- [x] have the values blink as described
- [x] connected wasm system simvars for valves and values
- [x] DOM now updates only when the values change

## Screenshots (if necessary)
![image](https://user-images.githubusercontent.com/43645128/96028900-368d2f80-0e5a-11eb-9993-28348ba9a465.png)
![image](https://user-images.githubusercontent.com/43645128/96032006-61798280-0e5e-11eb-91ce-c9bd8d8f0b98.png)


## References
https://www.youtube.com/watch?v=VUjfYWNijA8
![origin_press_page](https://user-images.githubusercontent.com/43645128/96028862-2b3a0400-0e5a-11eb-81f5-3ad4ca6e564f.png)


![image](https://user-images.githubusercontent.com/43645128/96023438-ba431e00-0e52-11eb-9fa8-d3f40cf2f6fa.png)
![image](https://user-images.githubusercontent.com/43645128/96023468-c7600d00-0e52-11eb-914c-1a49e7be224c.png)
![image](https://user-images.githubusercontent.com/43645128/96023488-cd55ee00-0e52-11eb-93b7-6d5ba5f95433.png)
![image](https://user-images.githubusercontent.com/43645128/96023524-dba40a00-0e52-11eb-9a57-b8176dff7db3.png)
![image](https://user-images.githubusercontent.com/43645128/96023537-dfd02780-0e52-11eb-9277-6bbc8276fbad.png)


## Additional context
In the end there should be a functioning ecam press page -- **currently waiting for the wasm system** 

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build.py will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, slick on the **Artifacts** drop down and click the **A32NX** link

noga-aviator commented:
![image](https://user-images.githubusercontent.com/35369933/96178423-ade0c300-0f2f-11eb-8ebc-10ba06b64ab5.png)
v/s seems to be stuck at 499 (you can see on the pfd that i'm more arround 4000)

nikotina91 commented:
yep, unfortunately all the values you see come from the game's simvars so I'm not sure it can be fixed yet, I could just do some math and show more real looking numbers but they'd just be "fake"

lukecologne commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_LowerECAM_PRESS.html:
```diff
@@ -0,0 +1,118 @@
+                <text id="sys1-text" style="visibility: hidden" class="st0p st9p" x="225" y="305">SYS  1</text>
+                <text id="sys2-text" class="st0p st9p" x="390" y="305">SYS  2</text>
+                <text id="lower-man-text" style="visibility: hidden" class="st0p st9p"  x="307.5" y="295">MAN</text>    
+                <text id="safety-text"class="st3p st9p" x="535" y="320">SAFETY</text>
```
```suggestion
                <text id="safety-text" class="st3p st9p" x="535" y="320">SAFETY</text>
```

lukecologne commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_LowerECAM_PRESS.html:
```diff
@@ -0,0 +1,118 @@
+                <text class="st3p st16p" x="72" y="168">0</text>
+                <text class="st3p st16p" x="112" y="130">8</text>
+                <text id="psi-gauge-int" class="st0p st9p"  x="128" y="170">8.</text>
+                <text id="psi-gauge-decimal"class="st0p st16p" style="font-size: 14px;" x="146" y="172">0</text>
```
```suggestion
                <text id="psi-gauge-decimal" class="st0p st16p" style="font-size: 14px;" x="146" y="172">0</text>
```

lukecologne reviewed (commented):
These missing spaces can cause issues

Benjozork commented:
Seems like the INLET and OUTLET labels are a bit closer to the line:

![image](https://user-images.githubusercontent.com/4503241/96180351-4b240300-0f00-11eb-97eb-3d318970d839.png)

Also, that V/S - FT/MIN indication is not aligned IRL. Same for CAB ALT - FT:

![image](https://user-images.githubusercontent.com/4503241/96180393-5d9e3c80-0f00-11eb-9959-1d94f31f4bd2.png)

nikotina91 commented:
I'll fix those right away, quick question, what do you think would be better, showing the simvar's values or, using documentation, showing some more realistic numbers that would only be related to that page until the wasm systems are in place?

Benjozork commented:
> what do you think would be better, showing the simvar's values or, using documentation, showing some more realistic numbers that would only be related to that page until the wasm systems are in place?

The latter.

Also, I see your are updating the DOM every update - please instead only update if the values have changed.

The way I do it is as follows:

```js
onUpdate() {
    this.updateValue()
}

updateValue() {
    const simvarValue = SimVar.GetSimVarValue("SOME SIMVAR:1", "number");

    if (simvarValue !== this.previousValue) {
        this.setValue(simvarValue);
    }
}

setValue(somethingElse) {
    this.something.innerText = somethingElse;
}

nikotina91 commented:
> > what do you think would be better, showing the simvar's values or, using documentation, showing some more realistic numbers that would only be related to that page until the wasm systems are in place?
> 
> The latter.
> 
> Also, I see your are updating the DOM every update - please instead only update if the values have changed.
> 
> The way I do it is as follows:
> 
> ```js
> onUpdate() {
>     this.updateValue()
> }
> 
> updateValue() {
>     const simvarValue = SimVar.GetSimVarValue("SOME SIMVAR:1", "number");
> 
>     if (simvarValue !== this.previousValue) {
>         this.setValue(simvarValue);
>     }
> }
> 
> setValue(somethingElse) {
>     this.something.innerText = somethingElse;
> }
> ```
page now:
![image](https://user-images.githubusercontent.com/43645128/96181686-b2f44100-0f34-11eb-9de4-11e882ddc385.png)

yeah I will definitely do that, makes much more sense, will have to wait until tomorrow though.

Benjozork commented:
Keep in mind that `console.log` lines should be removed when this is ready for review.

nikotina91 commented:
yeah, I will wait for the overhead panel pr to get through so I can have manual mode working as well, plus I'm having a little issue with the pid updating when the game is paused so it might take a bit before it's ready for review.

ImenesFBW commented:
Name: Imenes#8739
Date of testing: 19/10/2020
Version of the sim: 1950
PR Tested: #1373

Testing technique: Try-out
Steps made to test the Pull Request:

1. Loaded the plane in C&D and completed a 1h flight.
2. Observed the cabin pressure ECAM page for most of the flight.
3. After landing I played with manual pressure control and tried to take it to both extremes.

Overall Rating: Might need some small fixes. 
Conclusions of the testing: Working great and as documented in general, but my pack 1 and 2 triangles stayed amber throughout the flight even though they were functioning fine. After the flight I tried to play a bit with manual pressures settings and was not able to make the values of cabin pressure go red when they should. See the pics from the test below.

![111](https://user-images.githubusercontent.com/71195324/96400077-2ebdea00-11d0-11eb-9dcf-4ea12f972fbd.png)
![112](https://user-images.githubusercontent.com/71195324/96400089-35e4f800-11d0-11eb-8701-3ba29143e1b1.png)
![113](https://user-images.githubusercontent.com/71195324/96400098-3aa9ac00-11d0-11eb-94a0-df2f37a229a8.png)
![114](https://user-images.githubusercontent.com/71195324/96400117-42695080-11d0-11eb-8606-88eec6445ed8.png)
![115](https://user-images.githubusercontent.com/71195324/96400128-45644100-11d0-11eb-8795-88b8bae36fd3.png)

Great addition to the mod, thank you for your work!

nikotina91 commented:
thank you for testing! everything you mentioned should be fixed now.

Carter633 commented:
Name: British Avgeek YT -- Carter633 #1436
Date of testing: 19/10/2020
Version of the sim: 1.9.5.0
PR Tested: #1373

Testing technique: Full Flight from Cold and Dark - Genoa to Gatwick on VATSIM

Steps made to test the Pull Request:

1. Spawned Cold and Dark - during ground taxi checked Press page and left it on for duration of flight
2. Observed cabin pressure during climb and cruise
3. Observed press during descent phase and after landing

Overall Rating: Good
Conclusions of the testing: Seems to work well and as documented in detail above

ImenesFBW commented:
Thank you for the quick fixes @nikotina91 ! The triangles are now working correctly, I still can't seem too push the cabin altitude FT number to go red, when pushing it over with manual override.

![cabin](https://user-images.githubusercontent.com/71195324/96492163-86e20400-1243-11eb-9382-6c9e24d5a506.PNG)

nikotina91 commented:
that's interesting, they do go red for me with manual mode, I tested it a few times, were you on the ground or in the air? which state was the aircraft in? did you start from the runway or from cold and dark? @ImenesFBW

ImenesFBW commented:
Interesting maybe very specific to configuration. @nikotina91 the manual test was done after a full flight from C&D on a gate at the destination with both engines off and APU on.

nikotina91 commented:
I wasn't able to reproduce it (go figure), however I've changed the way the code sets the colors to the same used by the psi delta value (which, from what I can see from your pics, seems to change color correctly). @ImenesFBW 
Also I made it so that the indicators don't rotate past a certain point (values still update) and fixed the safety valve not closing (saw someone in discord talking about it).

ImenesFBW commented:
The indicator now indeed stops at max and stays there.

 For some reason the colors hate me. Still green. I tried with C&D and spawning on the runway, I also tried with the default livery and on a non-handcrafted airport, just in case. But in all cases still green, I'll wait and see if anyone else can reproduce it, or if its on my side. 

Are you testing with the version pulled via artifacts or your local version? Or I guess is your local version up to date with the latest master? @nikotina91

nikotina91 commented:
I was testing with the local version which I then copied into the repository folder and pushed to github, I really don't know what the problem could be, it's using the same function the psi value is using and that one works right? Anyway it's a bit late here and I'm kinda tired, will take a better look tomorrow morning (will test using the artifact too, you never know).

ImenesFBW commented:
Yeah that one goes amber. Really strange. Sorry to be a thorn in your side... 

Understandable, have a good night!

![last](https://user-images.githubusercontent.com/71195324/96516952-53fe3700-1268-11eb-8293-fa349c4de7e7.PNG)

camilo625 commented:
Name: Camilo625 #5575
Date of testing: Current date 19/10/2020
Version of the sim: 1.9.5.0
PR Tested: #1373 

Testing technique: Test cabin press while flying
Steps you've made to test the Pull Request:
1. Spawn C&D
2. Start a flight
3. Test cabin press while ascending
4. Test cabin press while descending

Overall Rating: To be improved
Conclusions of the testing: The page look good, the only thing i found its the same thing that was mentioned by a previous tester, about the number not going AMBER/RED.

![image](https://user-images.githubusercontent.com/38522306/96528075-227d6f80-1258-11eb-8fb8-4a9f24ecde3f.png)

FlyingCarpet7 commented:
Hi!
Noticed that the PACK 1 and PACK 2 text remains white with packs off and engines running. Should be amber.
Ref the FCOM picture above.

![Cab press](https://user-images.githubusercontent.com/71129337/96593978-6751e680-12ea-11eb-8d63-5a9c3463f961.png)

FlyingCarpet7 commented:
FL390. Delta P i bit low I believe. 
I did not see the delta P go above 8,2 but for some reason safety valve is open. When open it is not indicated AMBER as it should do.

![Cab press 3](https://user-images.githubusercontent.com/71129337/96598016-bbf76080-12ee-11eb-9c64-f15e9f7290ad.png)

lukecologne commented:
I wouldn't start perfecting the systems implementation. This will be done soon™ in C++, so spending time perfecting this one would be a waste of time IMHO. Just focus on the visual aspect in this PR.

nikotina91 commented:
oh yeah that one was just me testing last night and forgetting to put the value back to what they were, safety valve opened at 1.0 psi

Carter633 commented:
Name: British Avgeek YT -- Carter633 #1436
Date of testing: 20/10/2020
Version of the sim: 1.9.5.0
PR Tested: #1373

Testing technique: Full Flight from Cold and Dark - Heathrow to Malta on VATSIM
Steps made to test the Pull Request: same as previous test

1.Spawned Cold and Dark - during ground taxi checked 2.Press page and left it on for duration of flight
3.Observed cabin pressure during climb and cruise
4.Observed press during descent phase and after landing

Overall Rating: Good
Conclusions of the testing: normal operating is very good indeed. Ultimately if you use it as normal this works really well - in the sim there's no real need to force break everything so personally I think it looks and works very well indeed - naturally over time additional bits can be fixed and absolutely perfected outside normal operating, though that's my personal opinion

lukecologne commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_LowerECAM_CRZ.js:
```diff
@@ -153,40 +156,44 @@ var A320_Neo_LowerECAM_CRZ;
-                const cabinPressurePSI = cabinPressurePascal * Definitions.pascalToPSI;
-                const outsidePressurePSI = outsidePressureINHG * Definitions.inHgToPSI;
-                let pressureDiff = cabinPressurePSI - outsidePressurePSI;
+            const pressureDiff = SimVar.GetSimVarValue("L:A32NX_CABIN_PSI_DELTA", "PSI");
```
This can't be const, since it is reassigned later.

nikotina91 commented:
waiting on this until the wasm system is in.

Benjozork commented:
> waiting on this until the wasm system is in.

so this will stay draft until wasm system is merged, right ?

nikotina91 commented:
yeah, something like that

lukecologne commented:
Could you already change over to the WASM simvars? We would like to have your PR merged in order to properly be able test the WASM systems. If you want to see if your PR works with WASM, I could help you set it up.

nikotina91 commented:
I can, most of the work should already be done anyway, just need to know where to find the simvar names and the values they take.
I also have most of the work done for the bleed page if you need that too.

nikotina91 commented:
Didn't find any issue with 1.10.7

Benjozork commented:
What is the status on this ? Do we wait for the bleed system to be brought in ?

nikotina91 commented:
up to you, everything should be working, will most likely need a few minor changes to work well with the wasm system

Benjozork commented:
So, are the values faked currently ?

nikotina91 commented:
No they come from the wasm system simvars, someone asked me to switch to them above for the purpose of testing the system more easily when it's time to do so.
